### PR TITLE
add max steps

### PR DESCRIPTION
**Pull request by tkumor3**
Closes #

## Introduced changes

-

## Checklist

- [x] Linked relevant issue
- [ ] Updated relevant documentation
- [ ] Added relevant tests
- [x] Performed self-review of the code
- [ ] Added changes to `CHANGELOG.md`

### CODE REVIEW

File: crates/cheatnet/src/constants.rs
Comment (Arcticae) [inline crates/cheatnet/src/constants.rs:49]: What's the justification of this value?

**Review by Arcticae — CHANGES_REQUESTED** [review]
There's no e2e tests which should be ran here and docs

**Comment by tkumor3** [comment]
I need the new scarb version to write e2e tests. I asked Maciek to build nightly. Docs will be created in a separate PR. I assume that we want to inform people about it when the scarb 2.5.0 is released.

File: crates/cheatnet/src/constants.rs
Comment (tkumor3) [reply crates/cheatnet/src/constants.rs:49]: https://github.com/orgs/foundry-rs/projects/3/views/4?pane=issue&itemId=40320964

File: crates/forge/src/main.rs
Comment (piotmag769) [inline crates/forge/src/main.rs:82]: Space

File: crates/forge/src/main.rs
Comment (piotmag769) [inline crates/forge/src/main.rs:84]: I wonder if we want to have it configurable from the cli. Tbh I would even remove fuzzer config from the cli in the future, it's weird to have tests changing their internal behaviour depending on what cli args were used
cc @Arcticae

**Comment by piotmag769** [comment]
Tbh I think we can add docs in this PR and just add a temporary warning that the attribute in the cairo code needs scarb nightly-xyz to work

File: crates/cheatnet/src/constants.rs
Comment (MaksymilianDemitraszek) [reply crates/cheatnet/src/constants.rs:49]: Could you link the docs here?

File: .tool-versions
Comment (MaksymilianDemitraszek) [inline .tool-versions:1]: I assume we merge this one when 2.5.0 is released?

Comment (MaksymilianDemitraszek) [reply .tool-versions:1]: If not, we shouldn't merge this change to `.tool-versions`

File: crates/forge-runner/src/running.rs
Comment (MaksymilianDemitraszek) [inline crates/forge-runner/src/running.rs:295]: Please use match, it is safer beasue it checks for exhaustiveness.

File: crates/forge-runner/src/running.rs
Comment (MaksymilianDemitraszek) [inline crates/forge-runner/src/running.rs:296]: Does this function check if the value of max steps is correct? name suggests it.

File: crates/forge/src/main.rs
Comment (MaksymilianDemitraszek) [reply crates/forge/src/main.rs:84]: I assume @piotmag769 that this is usefull when all your tests have a lot of steps.

File: crates/forge-runner/src/running.rs
Comment (MaksymilianDemitraszek) [reply crates/forge-runner/src/running.rs:295]: Also please don't resolve it here. Please set the default value inside case in cli layer and make it not optional. This leaks cli logic into runner.

**Review by MaksymilianDemitraszek — CHANGES_REQUESTED** [review]
I don't think it works as supposed, can you use it to increase the maximum step limit inside vm/starknet 🤔 ? Also please include integration test checking if this works.

Please create an issue for docs
Please add integration test verifying the correctness when contract calls occur.

File: .tool-versions
Comment (piotmag769) [reply .tool-versions:1]: Why? setup scarb action detects .tool-versions and takes scarb version from it (yea, I got to know that yesterday, didn't expect it too xD)

Comment (MaksymilianDemitraszek) [reply .tool-versions:1]: Because it is unstable and we shouldn't run our tests on it.

Comment (piotmag769) [reply .tool-versions:1]: If we want to merge this one we have to run tests on this version. The other option is waiting with merging this PR until new stable scarb comes out

Comment (MaksymilianDemitraszek) [reply .tool-versions:1]: >If we want to merge this one we have to run tests on this version.

we should either ignore those tests or wait, we should not compromise existing features because of the new ones.

**Comment by tkumor3** [comment]
https://github.com/foundry-rs/starknet-foundry/pull/1384#pullrequestreview-1788587082

@MaksymilianDemitraszek I'm wondering, isn't it more useable to set VM max steps to starknet max value and return info about exceeding the limit?  Another thing, do we create only one instance of VM during test running?  we create `BlockContext` in 3 separate places. So I decided to get steps in the same way we get gas.

**Comment by piotmag769** [comment]
@tkumor3 yup, we create the VM in a lot of places. The only improvement I can see is keeping the current number of steps somewhere and passing the `steps_left` to each new VM, so it is handled automatically by blockifier. This way test gets executed to the point of exceeding max steps instead of executing the whole tests and asserting it wasn't exceeded at the end. I guess it is a more general question since we would want to handle available gas in a similar way
cc @war-in @Arcticae

**Comment by cptartur** [comment]
> @tkumor3 yup, we create the VM in a lot of places. The only improvement I can see is keeping the current number of steps somewhere and passing the `steps_left` to each new VM, so it is handled automatically by blockifier. This way test gets executed to the point of exceeding max steps instead of executing the whole tests and asserting it wasn't exceeded at the end. I guess it is a more general question since we would want to handle available gas in a similar way cc @war-in @Arcticae

Unsure about this one. Actually enforcing steps limit could be useful in cases where the users has possibly a long running test and wants it fail early. Though this argument is kind of a stretch.

Ultimately what matters is that we explicitly communicate to users that the steps limit was exceeded, probably by failing a test case.

**Comment by piotmag769** [comment]
@cptartur keep in mind we might want to include the current left resources info in the stack trace in the future, it will limit our ability to do so. This is actually the case with gas, check #1418

**Comment by cptartur** [comment]
> @cptartur keep in mind we might want to include the current left resources info in the stack trace in the future, it will limit our ability to do so. This is actually the case with gas, check #1418

If that's the case, and we need to do this properly for gas, we should handle steps in the same manner. So actually count and enforce it.

cc @MaksymilianDemitraszek @drknzz @Arcticae

File: crates/forge/tests/data/max_steps/tests/with_attr.cairo
Comment (Arcticae) [inline crates/forge/tests/data/max_steps/tests/with_attr.cairo:7]: That's kind of a bummer that we have to make it panicable by adding this assertion, since `max_steps` is kind of an assertion

File: crates/forge-runner/src/running.rs
Comment (Arcticae) [inline crates/forge-runner/src/running.rs:379]:
```suggestion
            bail!("Max steps limit exceeded. Limit: {max_steps}. Executed: {executed_steps}")
```

File: crates/forge-runner/src/running.rs
Comment (Arcticae) [reply crates/forge-runner/src/running.rs:296]: yeah the name seems off

File: crates/forge/tests/data/max_steps/tests/with_attr.cairo
Comment (Arcticae) [reply crates/forge/tests/data/max_steps/tests/with_attr.cairo:7]: Can we perhaps make a task to make it work like an assertion (so we don't require the panicable return type of test function)?

**Comment by Arcticae** [comment]
> If that's the case, and we need to do this properly for gas, we should handle steps in the same manner. So actually count and enforce it.

If it's possible to do via blockifier, then we should do it. Although i think gas is different since it has built in compilation mechanisms to generate code that "consumes" the gas before the function is run, so doing that might prove difficult. 

If it's not possible to be done automatically i think the current solution is reasonable.

File: crates/forge/tests/data/max_steps/tests/with_attr.cairo
Comment (piotmag769) [reply crates/forge/tests/data/max_steps/tests/with_attr.cairo:7]: Idk if it is worth doing and if it is doable, I believe it has to be panicable to get the `RunResult` in `run_test_case`, but maybe I'm wrong
cc @MaksymilianDemitraszek

**Comment by piotmag769** [comment]
When it comes to bumping `max_steps` above the starknet allowed maximum I think we can bump it to the value the user requested if it is greater than 3M, otherwise stay at 3M. We should check what happens if we exceed the `max_steps` limit from the blockifier context, I guess we would still fail but with different error message

File: crates/forge/tests/data/max_steps/tests/with_attr.cairo
Comment (MaksymilianDemitraszek) [reply crates/forge/tests/data/max_steps/tests/with_attr.cairo:7]: I believe it could be relaxed 🤔 not sure if this is worth the effort atm though. Could we create an issue?

**Comment by MaksymilianDemitraszek** [comment]
> I guess we would still fail but with different error message

@piotmag769 why with a different error message?

**Pull request closed without merging**
